### DISCUSSION TITLE
About action of invisible field (or dynamic field) in form

### DISCUSSION THREAD

**Discussion by jbee37142** [react-hook-form/react-hook-form#6759] [Q&A] [answered]
## TL;DR

1. Validation for fields that are not rendered and are not visible, is this intended?
2. `shouldUnregister: true` will be unregister field when unmount, but still live fields 😭 

### Code

https://codesandbox.io/s/react-hook-form-extensible-test-4n700?file=/src/App.js:1678-1694

```tsx
export default function App() {
  const { register, control, handleSubmit } = useForm({
    defaultValues: {
      isDeveloper: false
    }
  });
  const onSubmit = (data) => console.info(data);
  const onError = (error) => console.error(error);

  return (
    <form onSubmit={handleSubmit(onSubmit, onError)}>
      <BooleanRadio
        label="Are you Developer?"
        control={control}
        name="isDeveloper"
        rules={{ validate: (x) => x != null }}
        truthy={ /* render when value is true value */
          <label>
            <span>GitHub</span>
            <input
              {...register("social.github", {
                required: true,
                shouldUnregister: true
              })}
            />
          </label>
        }
      />
      <input type="submit" />
    </form>
  );
}
```

### Situation 1.

1. Click `<input type="submit" />`

#### Current Behavior

Check validate rules, even invisible fields.

```
// error
social: Object
  github: Object
    type: "required"
    message: ""
    ref: Object
```

#### Expected Behavior

Not validate, so submit works.

### Situation 2.

1. Click `yes` radio button
2. Write `social.github` input field - blahblah
3. Click `no` radio button
4.  Click `<input type="submit" />`

#### Current Behavior

fields -> `social.github` live! (registered status)

```ts
handleSubmit(data => console.log(data))
// isDeveloper: false
// social: {
//   github: 'blahblah'
// }

```

#### Expected Behavior

fields -> `social.github` removed

```ts
handleSubmit(data => console.log(data)) // isDeveloper: false
```

**Comment by bluebill1049** [comment, marked as answer]
![Screen Shot 2021-10-10 at 8 01 55 pm](https://user-images.githubusercontent.com/10513364/136689424-01bef536-2045-41b3-9681-509e3fd9701f.png)
I think you may miss this part of the doc, we will only be able to detected inputs at useForm hook with ref's callback, for external component, you will have to use invoke `unrgister` API

**Reply by bluebill1049** [reply]
```tsx
React.useEffect(() => {
  return () => unregister('inputname')
}, [unregister])
```

should fix your problem if i understand your issue correclty.

**Reply by jbee37142** [reply]
or, like this? 😭  :

```tsx
<BooleanRadio
  label="Are you Developer?"
  control={control}
  name="isDeveloper"
  rules={{ validate: (x) => x != null }}
  truthy={
    <GitHubInput
      control={control}
      name="social.github"
      rules={{ required: true }}
      shouldUnregister={true}
    />
  }
/>
```

**Reply by bluebill1049** [reply]
what if you just use `useController`? it's probably simpler? so you want input to appear base on the selection.

I would probably build them separately (separate of concern)

eg:

```tsx
const RadioButton = () => {
  return <Radios />
}

const ConditionalInput = () => {
  const value = useWatch({ name: 'radioButton' })

  if (value) {
    return <GitHubInput />
  }
  
  unregister('gitHubInput')

  return null;
}

const App = () => {
  return <div>
     <RadioButton />
     <ConditionalInput />
  </div>
}
```

**Reply by jbee37142** [reply]
@bluebill1049  thanks for good suggestion! (also, i agree `separate of concern` :) )

but, i don't want to use `useWatch` 😭  
first, I want to expose the dependencies of `RadioButton` and `GitHubInput`, but using `useWatch` makes to hide this dependency...!

And ...! my case have so  many condition...!

like this:

```tsx
<BooleanRadio
  control={control}
  name="A"
  truthy={
    <BooleanRadio
      control={control}
      name="B"
      truthy={<Input control={control} name="A+B" />}
      falsy={
        <BooleanRadio
          control={control}
          name="C"
          truthy={<Input control={control} name="A-B+C" />}
        />
      }
    />
  }
  falsy={
    <BooleanRadio
      control={control}
      name="D"
      truthy={<Input control={control} name="-A+D" />}
      falsy={<Input control={control} name="-A-D" />}
    />
  }
/>;
```

in this case... if using `useWatch`, `A-B+C` field have to watch `A`, `B`, `C` fields!

so... i want to encapsulation **boolean value** 😃 

my workaround is using `<Controller />`...!

**Reply by bluebill1049** [reply]
if you have some conditional logic, perhaps `useController` hook would make your code a bit clearer? at least the conditional is a bit more readable without the Component wrap with Compoennt? just some thoughts and suggestions :)

**Reply by jbee37142** [reply]
@bluebill1049 yeah, `useController` is also very nice!! But in this case, I thought it would be more readable to wrap all the fields with a `Controller` than to extract them as separate components!!
